Logger api

mborof opened the pull request:
This implements a logger API and makes changes to Propensity API so it be be integrated with logger later.

jpettitt commented on src/api/logger-api.js:
```diff
@@ -0,0 +1,252 @@
+/* eslint-disable no-unused-vars */
```
Why?

mborof replied:
Looks like its just an add-on complaining.  Removed.

jpettitt commented on src/api/propensity-api.js:
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable no-unused-vars */
```
Why?

mborof replied:
Looks like its just my add-on to visual studio code that's complaining.  Removed.

sohanirao commented on src/api/propensity-api.js:
```diff
@@ -263,38 +83,21 @@ export let PropensityScore;
+ * @extends {LoggerApi.LoggerApi}
  * @interface
  */
 export class PropensityApi {
```
Don't have to delete lines 277:to 297, you can simply at @override to lines 288 and 295

mborof replied:
Why re-write the same exact thing with an override?

sohanirao replied:
two reasons: 
1. completeness, when we extend interfaces, overriding existing interfaces is a common practice and it documents the new interface fully
2. looks like in this case even that might not work since, the argument to sendEvent() and sendSubscriptionState() now uses PropensityApi which is a clone to LoggerAPI but not the same. Does the closure allow you to send PropensityApi.Event when the method its overriding wants LoggerApi.Event?

mborof replied:
Yes because they are identical.  If I change one then it breaks though.

mborof replied:
I've added those functions back now.

sohanirao commented on src/api/propensity-api.js:
```diff
@@ -13,187 +13,7 @@
-   */
-  EVENT_CUSTOM: 'custom',
-}
+import * as LoggerApi from './logger-api';
```
Lets not delete this here, it serves as Propensity documentation

mborof replied:
Logger API must have event and SubscriptionState (currently defined in Propensity) in order to create functions that accept those enums.
Propensity must rely on Logger API so it can implement the interface declared in it
So if I do not move the primary definitions of Event and SubscriptionState to Logger API it creates a circular reference.

Redeclaring Event and SubscriptionState in Propensity is simply done so no one needs to change references to continue using SwG.

sohanirao replied:
I didn't mean create circular reference, I mean keep Propensity and Logger referes to Propensity API.

mborof replied:
Sorry, I left the word API out above.  Let me try to explain why you are asking for a circular reference again:  

1) PropensityAPI MUST reference LoggerAPI to implement its interface.  This cannot change.
2) LoggerAPI needs the enums currently defined in PropensityAPI because they are part of the interface it defines.  

If I do not move those enums out of Propensity API it is a circular reference.

sohanirao commented on src/api/propensity-api.js:
```diff
@@ -263,38 +83,21 @@ export let PropensityScore;
  */
 export let PropensityEvent;
 
+export const Event = LoggerApi.Event;
```
You would do the opposite. 
Like I have commented on the whole PR.

sohanirao commented:
Why not do it the other way around?

export PublisherEvent = PropensityApi.Event;
export SubscriptionState = PropensityApi.SubscriptionState;

mborof commented:
1) Logger API must have event and SubscriptionState (currently defined in Propensity) in order to create functions that accept those enums.
2) Propensity must rely on Logger API so it can implement the interface declared in it

So if I do not move the primary definitions of Event and SubscriptionState to Logger API it creates a circular reference.

Redeclaring Event and SubscriptionState in Propensity is simply done so no one needs to change references to continue using SwG.

mborof commented:
But Propensity API needs to reference Logger API so it can implement its
interface.

On Thu, Jul 18, 2019 at 2:37 PM sohanirao <notifications@github.com> wrote:

> *@sohanirao* commented on this pull request.
> ------------------------------
>
> In src/api/propensity-api.js
> <https://github.com/subscriptions-project/swg-js/pull/588#discussion_r305125484>
> :
>
> > -   * event, must be set to true or false depending on if the event
> -   * was generated as a result of a user action.
> -   * The JSON block can provide the event name for the custom event.
> -   * For example:
> -   * const propensityEvent = {
> -   *   name: 'custom',
> -   *   active: true,
> -   *   data: {
> -   *     'event_name': 'social_share',
> -   *     'platform_used': 'whatsapp'
> -   *   }
> -   *  }
> -   */
> -  EVENT_CUSTOM: 'custom',
> -}
> +import * as LoggerApi from './logger-api';
>
> I didn't mean create circular reference, I mean keep Propensity and Logger
> referes to Propensity API.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/subscriptions-project/swg-js/pull/588?email_source=notifications&email_token=ALZ5325YU7WKC2QGMUU6GF3QADPBRA5CNFSM4IEIELI2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB65R4HQ#discussion_r305125484>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALZ5322SSULLR5NHT2SOQHDQADPBRANCNFSM4IEIELIQ>
> .
>

mborof commented:
I've done 2 things:
  1) Removed the repeating of the logging definitions from propensity-api.  It is better without them.
  2) Added comments telling publishers to look in logger-api for the logging enums and object defs.

sohanirao commented on src/api/propensity-api.js:
```diff
@@ -287,38 +110,28 @@ export let PropensityScore;
  *
  *  @typedef {{
- *    name: string,
+ *    name: !Event,
```
this might be a problem since publishers dont have access to enum types. a good way to confirm is try using this api is scenic demo.
